File: golang/generate_signature.go
package generate_signature

import (
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/pem"
	"errors"
	"fmt"
	"github.com/google/uuid"
	"time"
)

const privateKey = `your private key`

func GenerateSignatureHeader(appId, method, path, payload string) (string, error) {
	timestamp := time.Now().UTC().Unix()
	nonceStr := uuid.New().String()

	message := fmt.Sprintf("%s\n%s\n%d\n%s\n%s", method, path, timestamp, nonceStr, payload)
	signature, err := Sign([]byte(message), []byte(privateKey))
	if err != nil {
		return "", err
	}

	return fmt.Sprintf("TAMS-SHA256-RSA app_id=%s,timestamp=%d,nonce_str=%s,signature=%s", appId, timestamp, nonceStr, base64.StdEncoding.EncodeToString(signature)), nil
}

func Sign(message []byte, privateKeyBytes []byte) ([]byte, error) {
	block, _ := pem.Decode(privateKeyBytes)
	if block == nil {
		return nil, errors.New("invalid private key")
	}

	privateKey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
	if err != nil {
		return nil, err
	}

	hashed := sha256.Sum256(message)

	signature, err := rsa.SignPKCS1v15(rand.Reader, privateKey, crypto.SHA256, hashed[:])
	if err != nil {
		return nil, err
	}

	return signature, nil
}

File: nodejs/generate_signature.js
import { createHash, createSign } from "crypto";
import { readFileSync } from "fs";

function generate_signature(method, url, body, app_id, private_key_path) {
  const method_str = method.toUpperCase();
  const url_str = url;
  const timestamp = Math.floor(Date.now() / 1000).toString();
  const nonce_str = createHash('md5').update(timestamp).digest("hex");
  const body_str = body || "";
  const to_sign = `${method_str}\n${url_str}\n${timestamp}\n${nonce_str}\n${body_str}`;
  
  const private_key = readFileSync(private_key_path);
  const sign = createSign('RSA-SHA256');
  sign.update(to_sign);
  const signature = sign.sign(private_key, "base64");
  const auth_header = `TAMS-SHA256-RSA app_id=${app_id},nonce_str=${nonce_str},timestamp=${timestamp},signature=${signature}`;
  return auth_header;
}

